### PR TITLE
Support host user namespace 

### PR DESCRIPTION
**Pull request by pweil-**
@smarterclayton @ncdc @mrunalp This adds the host user flag to the PSC and threads through the parameter for all containers of the pod (including the pause container).  This does not include any PSP changes.

Also, to see the UsernsMode in inspect output you need a compatible docker version however after testing on 1.10.3 and 1.12 there didn't seem to be any adverse affects to having the parameter set in the api on the lower version. 

```
[pweil@localhost userns]$ docker ps
CONTAINER ID        IMAGE                                      COMMAND                  CREATED             STATUS              PORTS               NAMES
9c4a8f0a2b7b        nginx                                      "nginx -g 'daemon off"   57 seconds ago      Up 55 seconds                           k8s_nginx.b2db2f1d_userns_default_075cf376-63bc-11e6-9447-54ee752009cb_263fdfb4
9dd4c7ee1d34        gcr.io/google_containers/pause-amd64:3.0   "/pause"                 58 seconds ago      Up 57 seconds                           k8s_POD.b2390301_userns_default_075cf376-63bc-11e6-9447-54ee752009cb_de217910
[pweil@localhost userns]$ docker inspect 9dd4 | grep UsernsMode
            "UsernsMode": "host",
[pweil@localhost userns]$ docker inspect 9c4 | grep UsernsMode
            "UsernsMode": "host",
[pweil@localhost userns]$ cat pod.yaml
# Copy of pod.yaml without file extension for test
apiVersion: v1
kind: Pod
metadata:
  name: userns
  labels:
    name: userns
spec:
  securityContext:
      hostUser: true
  containers:
  - name: nginx
    image: nginx
    ports:
    - containerPort: 80
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30684)

### CODE REVIEW

**Comment by thockin** [comment]
I am fine with this.  Do we have guidance for why someone would want to do this?

**Comment by smarterclayton** [comment]
Use case:

As an administrator I want to secure my cluster by default (by using the
userns=remap option on Docker 1.10+), but I still want to run pods on the
cluster that allow me to use host volumes.  Since setting userns=remap
changes the behavior of Kube hostPath (container user would be 50, and not
have access), the admin needs a way to escape the user namespace.  Since
user namespace dominates the other namespaces, in Docker specifying
IPC=host or net=host has the implication that userns=host, but if we didn't
expose userns=host to the admin explicitly there would be no way to have
compatible behavior.

We had considered suggesting that hostPath *automatically* escape userNS,
but that felt far more scary as a security posture.  Since we have a number
of other "host namespace" flags, this is at least consistent, and it's up
to the admin to know that they have to use it.  I believe this is
compatible with the path we would take with Rkt in the future - Rkt can
*automatically* remap (if the admin configures it), and then can respect
userns=host and bypass that.

**Comment by pweil-** [comment]
@k8s-bot test this issue #30683

**Comment by smarterclayton** [comment]
Is PSP changes separate?

**Comment by pweil-** [comment]
> Is PSP changes separate?

Yes

**Comment by ncdc** [comment]
cc @kubernetes/sig-node

**Comment by pmorie** [comment]
What's the rationale for defaulting to HostUser defaulting to false?

**Comment by pweil-** [comment]
> What's the rationale for defaulting to HostUser defaulting to false?

All host namespace usage defaults to false (net, ipc, pid) and requires you to explicitly break out.

**Comment by pmorie** [comment]
@pweil- I get that -- but the defaults for those are also in line with existing behaviors.  Remap by default is a pretty big change, agree?

**Comment by pweil-** [comment]
The remap is done by docker daemon arguments (ie `ExecStart=/usr/bin/docker daemon --userns-remap=`).  This is falling in line with the current behavior for `HostConfig` which would default `usernsMode=""` it's just allowing you to set the value to `"host"` now.

**Comment by smarterclayton** [comment]
I believe this is the rkt behavior today.

**Comment by vishh** [comment]
+1 for not performing a remap by default.

**Comment by mrunalp** [comment]
The remap is at the daemon level so the onus is on whatever is launching the docker daemon to remap or not.

**Comment by mrunalp** [comment]
If one were to choose to enable user namespaces with docker, then containers that need to access host namespaces break unless this option is provided. This is because all other namespaces are owned by user namespaces. So, for e.g if there is a pod that needs to listen on low port on the host network namespace it can only do that if it also uses the host user namespace. This PR enables that.

**Comment by pweil-** [comment]
@k8s-bot test this issue #30683

**Comment by pweil-** [comment]
@k8s-bot test this issue #27503

**Comment by thockin** [comment]
If it is true that hostNetwork and hostIPC are broken without this (when
user has UserNS enabled) then shouldn't we just enable this automatically
when the other host* are set?  Otherwise it is a very weird failure mode
and we force users to know the details of linux namespaces

**Comment by smarterclayton** [comment]
Today in docker saying --ipc=host --net=host actually disables user
namespace.  But volume mounting a host path does not.

**Comment by mrunalp** [comment]
I just checked latest docker to be certain. It is in fact blocking --net=host, --pid=host and --ipc=host when user namespaces are enabled. It should be allowed when --userns=host. I'll create a PR for that. 

But yeah, setting --userns=host when other namespaces are using the host namespaces should be a good default unless you have edge cases like wanting to bind to higher port on the host while still using user namespaces in the container.

**Comment by thockin** [comment]
On Tue, Aug 16, 2016 at 2:11 PM, Mrunal Patel <notifications@github.com> wrote:
>
> I just checked latest docker to be certain. It is in fact blocking --net=host, --pid=host and --ipc=host when user namespaces are enabled. It should be allowed when --userns=host. I'll create a PR for that.

Sorry, I have trouble parsing this.What is "it" here?

> But yeah, setting --userns=host when other namespaces are using the host namespaces should be a good default unless you have edge cases like wanting to bind to higher port on the host while still using user namespaces in the container.

Can we write out a full table of the intersection of these features,
what features should imply other features, and what features should
block other features?

**Comment by mrunalp** [comment]
Created PR https://github.com/docker/docker/pull/25771

@thockin --pid, --net and --ipc should be allowed to be set to `host` when --userns is set to `host`. 
I'll try to come up with a table.

**Comment by mrunalp** [comment]
Assuming that the docker daemon is started with --userns-map=default, here are use cases and how to achieve them:


|  Use case | Support  |
|---|---|
|  Load a kernel module | --userns=host  |
|  Create a device using mknod |  --userns=host |
|  Listen to a low port on host | --userns=host, --net=host* |
|  Monitor processes on host |  --userns=host, --pid=host* |
|  Listen to high port on host | --net=host, --userns=container_private^|
| share net,pid namespace between containers| --net,--pid=container:id implicitly shares userns as well|

[*] = requires https://github.com/docker/docker/pull/25771
[^] = not supported in docker but could be done using runc

Let me know if I missed any and I will update the table for reference.

**Comment by mrunalp** [comment]
I have edited the comment with the table, so please take a look at it on github as you probably didn't get my updates.

**Comment by smarterclayton** [comment]
Add to table "access a host path as the root user", i.e. "monitoring daemon that needs to look at /proc"

**Comment by mrunalp** [comment]
@smarterclayton done.

**Comment by mrunalp** [comment]
https://github.com/docker/docker/pull/25771 has been merged.

**Comment by euank** [comment]
It looks like there are enough intricacies, caveats, and areas that this touches that it's worth creating a quick design/proposal document explaining the options and why this is the right one and what the future support might look like.

At the very least, it would be a good thing to point people at rather than @mrunalp's comments in this issue for that table.

Also, we need a `release-note` label somewhere for a new API thing I think.

**Comment by pweil-** [comment]
@euank @smarterclayton @mrunalp @thockin @ncdc added proposal commit.  PTAL

File: docs/proposals/user-namespace.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+1.  There is currently a bug in Docker that does not allow you to use
+`--net=host`, `--pid=host`, or `--ipc=host` when remapping is enabled.
+This is fixed in https://github.com/docker/docker/pull/25771 such that
+if you are enabling user namespace remapping you may use `--net=host`, `--pid=host`, 
+and `--ipc=host` flags if `--userns=host`.
+
+## Use Cases
+
+Assuming that the docker daemon is started with --userns-remap=default,
```
Comment (smarterclayton) [inline docs/proposals/user-namespace.md]: You should start this with "Kubernetes cluster administrators wish to have a higher level of default cluster security on Docker 1.10+ by enabling user namespace remapping, but want to be able to continue to run cluster level actions for monitoring."

Comment (pweil-) [inline docs/proposals/user-namespace.md]: added

File: docs/proposals/user-namespace.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    Cons
+    *  Requires extra flags/config - this defaulting should only occur
+    when remapping is configured on the Docker daemon.
+    *  Does not support edge cases like using user namespaces but binding
+    to high ports on the host.
+
+## Proposal
+
+For 1.4 we can add the `HostUser` flag to provide immediate support
+for anyone who has a use case that is not currently supported.  We can
```
Comment (smarterclayton) [inline docs/proposals/user-namespace.md]: Add "... and enable users on Docker 1.10+ to test and enable userns=remap"

File: docs/proposals/user-namespace.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+2.  Provide a `HostUser` flag and provide auto setting.
+
+    Pros
+    *  System does the right thing when it can
+
+    Cons
+    *  Requires extra flags/config - this defaulting should only occur
+    when remapping is configured on the Docker daemon.
+    *  Does not support edge cases like using user namespaces but binding
+    to high ports on the host.
```
Comment (timstclair) [inline docs/proposals/user-namespace.md]: Also: User might accidentally cause a container to be run in the host user ns. I'm generally weary of implicitly elevating privileges.

File: docs/proposals/user-namespace.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+## Options
+
+1.  Provide a `HostUser` flag and require manual setting.
+
+    Pros
+    *  Quick and easy
+    *  Allows full flexibility
+
+    Cons
+    *  Requires linux specific knowledge
```
Comment (timstclair) [inline docs/proposals/user-namespace.md]: I think this could be mitigated with good error messages in validation.

**Comment by thockin** [comment]
What I wanted to see was something like:

--net=host, --pid=host --ipc=host, --user=host, hostPath volume = true
hostPath

--net=default, --pid=default --ipc=default, --user=default, hostPath volume
= ???

--net=default, --pid=default --ipc=default, --user=host, hostPath volume =
???

etc


I can not map the matrix of these in my head.  If, as you suggest,
--user=host implies --ipc=host, should we fail if the user didn't ask for
all of them (POLS) or should we assume that --user=host changes --ipc and
friends without warning?

We have 4 flags and I think we need to see all 16 combinations of the flags
explored for what works and what explodes and what does something
unexpected.

CF the table I did comparing docker CMD and ENTRYPOINT.  It revealed a few
really weird behaviors.

File: docs/proposals/user-namespace.md (hunk 4 above)
Comment (pweil-) [inline docs/proposals/user-namespace.md]: Agree though I think, as with defaulting, the validation would be reliant on knowing that an external process was started with the remap option.  I would suggest that if you're already setting flags for any other host namespace you probably know why you're doing it and have at least some knowledge of other namespace requirements.

File: docs/proposals/user-namespace.md (hunk 3 above)
Comment (pweil-) [inline docs/proposals/user-namespace.md]: added

**Comment by pweil-** [comment]
@thockin For our purposes of when to use host user namespace we can boil it down to the following rules:

1.  When setting any other namespace flag to `host`
    * Not strictly required for `net=host` if not binding to low ports 
     but if the system provides defaulting it covers the 80% case
2.  When using capabilities that aren't tied to any namespace
    * MKNOD
    * SYS_MODULE
    * SYS_TIME
    * SYS_ADMIN
    * others?  (@mrunalp is looking at source to determine this)
3.  When explicitly granted permissions to use a host volume and 
escape the private user namespace

I've added this to the proposal.  We can provide sane defaulting for cases 1 & 2 above if we have a way to know that user remapping is enabled in the docker daemon.  For the case of explicit permission it would be up to the user to request to use the host user namespace.

**Comment by pweil-** [comment]
@k8s-bot test issue #ignore

(errors checking out the branch)

File: docs/proposals/user-namespace.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+1.  When setting any other namespace flag to `host`
+    * Not strictly required for `net=host` if not binding to low ports
+     but if the system provides defaulting it covers the 80% case
+2.  When using capabilities that aren't tied to any namespace
+    * MKNOD
+    * SYS_MODULE
+    * SYS_TIME
+    * others?  (@mrunalp is looking at source to determine this)
+3.  When explicitly granted permissions to use a host volume and
```
Comment (vishh) [inline docs/proposals/user-namespace.md]: What do you mean by `explicitly granted permissions`? Any pod can use a host volume today. Are you implying that using host volumes will automatically turn off userns?

File: docs/proposals/user-namespace.md (hunk 4 above)
Comment (vishh) [inline docs/proposals/user-namespace.md]: Isn't this already the case with so many features (e.x. SELinux) in the API?

File: docs/proposals/user-namespace.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+    Pros
+    *  System does the right thing when it can
+
+    Cons
+    *  Requires extra flags/config - this defaulting should only occur
+    when remapping is configured on the Docker daemon.
+    *  Does not support edge cases like using user namespaces but binding
+    to high ports on the host.
+    *  The user may unexpectedly run a container with the host user namespace.
```
Comment (vishh) [inline docs/proposals/user-namespace.md]: Also, admins have to be aware of the fact that certain capabilities result in lack of userns.

File: docs/proposals/user-namespace.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+## Proposal
+
+For 1.4 we can add the `HostUser` flag to provide immediate support
```
Comment (vishh) [inline docs/proposals/user-namespace.md]: Why would this go in v1.4 given that https://github.com/docker/docker/pull/25771 just got merged?

File: docs/proposals/user-namespace.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+
+## Additional Concerns
+
+When any new field is added to the `SecurityContext` it should also be
+added to the `PodSecurityPolicy` so that it may be controlled by an
+administrator.
+
+In this case the field is a boolean and should be controlled by an option
+that allows it to be set or not.  PSP should not provide a default.
```
Comment (vishh) [inline docs/proposals/user-namespace.md]: PSP typically allows all operations right? So disabling userns would be the default behavior right?

File: docs/proposals/user-namespace.md (hunk 7 above)
Comment (vishh) [inline docs/proposals/user-namespace.md]: A flag or a field in the API?

File: docs/proposals/user-namespace.md (hunk 5 above)
Comment (smarterclayton) [inline docs/proposals/user-namespace.md]: What do you mean by explicitly granted permissions? Any pod can use a host
volume today. Are you implying that using host volumes will automatically
turn off userns?


PodSecurityPolicy controls whether you can use host volumes, since host
volumes can trivially root a cluster.

I don't thing hostVolume should turn off user ns implicitly, because there
are use cases for userns=on+hostVolume that are secure.

File: docs/proposals/user-namespace.md (hunk 4 above)
Comment (smarterclayton) [inline docs/proposals/user-namespace.md]: Isn't this already the case with so many features (e.x. SELinux) in the API?


Yes, just making sure we explicitly call it out each time.

File: docs/proposals/user-namespace.md (hunk 7 above)
Comment (smarterclayton) [inline docs/proposals/user-namespace.md]: Why would this go in v1.4 given that docker/docker#25771
<https://github.com/docker/docker/pull/25771> just got merged?


Red Hat is planning to backport to Docker 1.10 so that customers can get
better security from Kube clusters by default.

File: docs/proposals/user-namespace.md (hunk 8 above)
Comment (smarterclayton) [inline docs/proposals/user-namespace.md]: > +For 1.4 we can add the `HostUser` flag to provide immediate support
> +for anyone who has a use case that is not currently supported and
> +enable users on Docker 1.10+ to test and enable `userns=remap`.  We can
> +then iterate on smart defaulting in future releases to ensure that the
> +system will do the right thing when remapping is known to be configured.
> +
> +
> +## Additional Concerns
> +
> +When any new field is added to the `SecurityContext` it should also be
> +added to the `PodSecurityPolicy` so that it may be controlled by an
> +administrator.
> +
> +In this case the field is a boolean and should be controlled by an option
> +that allows it to be set or not.  PSP should not provide a default.

PSP typically allows all operations right? So disabling userns would be the
default behavior right?


It does in Kube, but I think that might be wrong in general.  OpenShift
defaults to safe by default by putting all restrictions on unless you
escalate - I think we want that for Kube eventually.  We could argue that
defaulting on makes sense in the immediate timeframe for Kube, but since
you have to turn userns on also, it seems reasonable to default to off in
the API and let the Kubelet/runtime provide that explicit behavior.

Comment (thockin) [inline docs/proposals/user-namespace.md]: I've said my peace.  I STILL don't feel like I have my head wrapped around
it, so I fully expect this to suck for users, too.  It's pretty niche, I
hope

File: docs/proposals/user-namespace.md (hunk 5 above)
Comment (pweil-) [inline docs/proposals/user-namespace.md]: > Are you implying that using host volumes will automatically turn off userns

No.  This case is covering any niche cases where you have permissions to use a PSP that allows you to escape user ns and also use host volumes.  This would be an explicit request from the user to enter that state though, not something we could reasonably default.  Items 1 and 2 could be defaulted.

File: docs/proposals/user-namespace.md (hunk 7 above)
Comment (pweil-) [inline docs/proposals/user-namespace.md]: > A flag or a field in the API?

Right, we currently have no way to express that a container should use the host user namespace and have that passed to docker like we do with the other namespace flags.  PSP control and DTRT defaulting aside, this should be an option in the security context IMO.

File: docs/proposals/user-namespace.md (hunk 6 above)
Comment (pweil-) [inline docs/proposals/user-namespace.md]: added

**Comment by smarterclayton** [comment]
By analogy, a container today defaults to being in net namespace.  That default is provided by the container runtime.  A container runtime could choose to be in a configuration with no network isolation.  In that case, hostNetwork is ignored, rather than triggering a failure to submit a pod with hostNetwork: false.  At some level, the interaction between container security defaults and the user input probably has to be predictable (I must have hostX) but leave the default up to the CR.

**Comment by vishh** [comment]
Given the time crunch, can we limit the scope of this feature in v1.4 to just provide an experimental compatibility mode in kubelet? If a user wants to run with userns in docker, then kubelet will automatically set userns to native whenever a pod matches one of the criteria mentioned in https://github.com/kubernetes/kubernetes/pull/30684#issuecomment-240810269?

**Comment by smarterclayton** [comment]
The other two don't bother me that much, but hostVolume worries me
slightly.  I agree the time crunch is an issue.

**Comment by pweil-** [comment]
> Given the time crunch, can we limit the scope of this feature in v1.4 to just provide an experimental compatibility mode in kubelet?

ok, let me give this a go.  Will submit something asap

**Comment by vishh** [comment]
@pweil- Can you summarize the implementation plan for v1.4?

**Comment by pweil-** [comment]
Implementation plan for 1.4

1.  add experimental flag to the kubelet to the effect of `experimental-host-user-namespace-defaulting` which should only be enabled when docker daemon remapping is enabled AND the experimental behavior is desired
2.  pass said flag to `DockerMangager`
3.  when said flag is enabled and a pod falls into one of these conditions set `UsernsMode: host` in `DockerMangager`
    1.  any other host namespace is used
    1.  mknod, sys_module, or sys_time caps are used
    1.  a host volume is used

Will open as a separate PR and also create a feature request for a possible api change in 1.5.

**Comment by smarterclayton** [comment]
Just to clarify, was there any argument so far that hostUser is
inconsistent with our existing pattern?  I just thought the concern was
about how it would map to admin expectations and the "if X then Y" behavior
for IPC/net?

**Comment by thockin** [comment]
I think there is ample precedent of hostUser as an API, but this is so much
more delicate than the others, I don't want to feel rushed.

**Comment by smarterclayton** [comment]
Yeah, I'm not advocating for rushing the API, just want clarity on concerns
so they are properly reflected in the 1.5 proposal.

**Comment by pweil-** [comment]
Proposed flag PR: https://github.com/kubernetes/kubernetes/pull/31169

File: docs/proposals/user-namespace.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+| Share net,pid namespaces between containers  | --net,--pid=container:id implicitly shares userns as well  |
+
+We can make a couple of generalizations from the table:
+
+1. Certain capabilities that aren't tied to any namespace such as
+CAP_MKNOD, CAP_SYS_MODULE or CAP_SYS_TIME are requested, then they
+imply userns=host.
+2. Setting any other namespace to host could imply userns=host.
+
+[*] = requires docker/docker#25771
```
Comment (euank) [inline docs/proposals/user-namespace.md]: this * should be escaped.

**Comment by euank** [comment]
I think the current proposal doesn't specify a number of details clearly enough.

1) It doesn't go into enough detail in runtime support and failure modes IMO. The fact that this requires a very new version of docker matters and what happens when your runtime can't support this also matters.
2) This doesn't address runtimes other than Docker at all. It's worth giving a quick moment to "under rkt ..."
3) This doesn't talk about how this will be modeled in CRI or what will happen for e.g. hypervisor based runtimes.

Since it turns out this is a 1.5 feature, not 1.4 feature, I think we can step back and flesh out the details of what the user-facing API should look like a little bit more as well.

**Comment by pweil-** [comment]
closing this in favor of a full feature proposal and experimental support in https://github.com/kubernetes/kubernetes/pull/31169
